src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs

```diff
@@ -15,129 +16,117 @@ namespace Microsoft.CodeAnalysis
-                    if (_text == null)
-                    {
-                        return InitRecoverable(await _initialSource!.GetValueAsync(cancellationToken).ConfigureAwait(false));
-                    }
-                }
+                // replace initial source with recovarable text, if it hasn't been replaced already:
+                Interlocked.CompareExchange(
+                    ref _initialSourceOrRecoverableText,
+                    value: new RecoverableText(await source.GetValueAsync(cancellationToken).ConfigureAwait(false), _services),
+                    comparand: source);
             }
 
-            var text = await _text.GetValueAsync(cancellationToken).ConfigureAwait(false);
-            return TextAndVersion.Create(text, _version, _filePath, _loadDiagnostic);
+            var recoverableText = (RecoverableText)_initialSourceOrRecoverableText;
+            return recoverableText.ToTextAndVersion(await recoverableText.GetValueAsync(cancellationToken).ConfigureAwait(false));
         }
 
-        private TextAndVersion InitRecoverable(TextAndVersion textAndVersion)
+        private sealed class RecoverableText : WeaklyCachedRecoverableValueSource<SourceText>, ITextVersionable
         {
-            _initialSource = null;
-            _version = textAndVersion.Version;
-#pragma warning disable CS0618 // Type or member is obsolete
-            _filePath = textAndVersion.FilePath;
-#pragma warning restore
-            _loadDiagnostic = textAndVersion.LoadDiagnostic;
-            _text = new RecoverableText(this, textAndVersion.Text);
-            _text.GetValue(CancellationToken.None); // force access to trigger save
-            return textAndVersion;
-        }
+            private readonly ITemporaryStorageServiceInternal _storageService;
+            public readonly VersionStamp Version;
+            public readonly Diagnostic? LoadDiagnostic;
 
-        private sealed class RecoverableText : WeaklyCachedRecoverableValueSource<SourceText>
-        {
-            private readonly RecoverableTextAndVersion _parent;
-            private ITemporaryTextStorageInternal? _storage;
+            public ITemporaryTextStorageInternal? _storage;
 
-            public RecoverableText(RecoverableTextAndVersion parent, SourceText text)
-                : base(new ConstantValueSource<SourceText>(text))
+            public RecoverableText(TextAndVersion textAndVersion, HostWorkspaceServices services)
+                : base(new ConstantValueSource<SourceText>(textAndVersion.Text))
             {
-                _parent = parent;
+                _storageService = services.GetRequiredService<ITemporaryStorageServiceInternal>();
+
+                Version = textAndVersion.Version;
+                LoadDiagnostic = textAndVersion.LoadDiagnostic;
             }
 
+            public TextAndVersion ToTextAndVersion(SourceText text)
+                => TextAndVersion.Create(text, Version, filePath: null, LoadDiagnostic);
```

Comment:
ok that the filepath is null? seems... concerning.

Reply:
the path is unused (obsolete)